lib/Driver/UnixToolChains.cpp

```diff
@@ -278,6 +278,12 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
   if (job.getKind() == LinkKind::Executable && context.OI.SelectedSanitizers) {
     Arguments.push_back(context.Args.MakeArgString(
         "-fsanitize=" + getSanitizerList(context.OI.SelectedSanitizers)));
+
+    // The TSan runtime depends on the blocks runtime and libdispatch.
+    if (context.OI.SelectedSanitizers & SanitizerKind::Thread) {
+      Arguments.push_back("-lBlocksRuntime");
```

Comment:
This is slightly worrisome.  `-fblocks-runtime-optional` controls where you have a blocks runtime library or not that you link against.  (The GNU ObjC runtime also provides the blocks runtime implicitly).  I wish that we could delegate to the clang driver for the blocks runtime linkage.

Reply:
> [...] delegate to the clang driver for the blocks runtime linkage.

This is what I am trying to do here. I don't really understand what you mean. What should we do instead?